src/V3DfgSynthesize.cpp

```diff
@@ -0,0 +1,1699 @@
+        struct Assignment final {
+            DfgVertexSplice* m_lhsp;
+            uint32_t m_idx;
+            DfgVertex* m_rhsp;
+            Assignment() = delete;
+            Assignment(DfgVertexSplice* lhsp, uint32_t idx, DfgVertex* rhsp)
+                : m_lhsp{lhsp}
+                , m_idx{idx}
+                , m_rhsp{rhsp} {}
+        };
+
+        // Convert each concatenation LHS separately, gather all assignments
+        // we need to do into 'assignments', return true if all LValues
+        // converted successfully.
+        std::vector<Assignment> assignments;
+        const std::function<bool(AstNodeExpr*, DfgVertex*)> convertAllLValues
+            = [&](AstNodeExpr* lhsp, DfgVertex* vtxp) -> bool {
+            // Simplify the LHS, to get rid of things like SEL(CONCAT(_, _), _)
+            lhsp = VN_AS(V3Const::constifyExpensiveEdit(lhsp), NodeExpr);
+
+            // Concatenation on the LHS, convert each parts
+            if (AstConcat* const concatp = VN_CAST(lhsp, Concat)) {
+                AstNodeExpr* const cLhsp = concatp->lhsp();
+                AstNodeExpr* const cRhsp = concatp->rhsp();
+                // Convert Left of concat
+                FileLine* const lFlp = cLhsp->fileline();
+                AstNodeDType* const lDtp = DfgGraph::toDfgDType(cLhsp->dtypep());
+                DfgSel* const lVtxp = make<DfgSel>(lFlp, lDtp);
+                lVtxp->fromp(vtxp);
+                lVtxp->lsb(cRhsp->width());
+                if (!convertAllLValues(cLhsp, lVtxp)) return false;
+                // Convert Rigth of concat
+                FileLine* const rFlp = cRhsp->fileline();
+                AstNodeDType* const rDtp = DfgGraph::toDfgDType(cRhsp->dtypep());
+                DfgSel* const rVtxp = make<DfgSel>(rFlp, rDtp);
+                rVtxp->fromp(vtxp);
+                rVtxp->lsb(0);
+                return convertAllLValues(cRhsp, rVtxp);
+            }
+
+            // Non-concatenation, convert the LValue
+            const auto pair = convertLValue(lhsp);
+            if (!pair.first) return false;
+            assignments.emplace_back(pair.first, pair.second, vtxp);
+            return true;
+        };
+
+        // Convert the given LHS assignment, give up if any LValues failed to convert
+        if (!convertAllLValues(lhsp, vtxp)) return false;
+
+        // All successful, connect the drivers
+        for (const Assignment& item : assignments) {
+            if (DfgSplicePacked* const spp = item.m_lhsp->template cast<DfgSplicePacked>()) {
+                spp->addDriver(flp, item.m_idx, item.m_rhsp);
+            } else if (DfgSpliceArray* const sap = item.m_lhsp->template cast<DfgSpliceArray>()) {
+                AstUnpackArrayDType* const lDtp = VN_AS(sap->dtypep(), UnpackArrayDType);
+                AstNodeDType* const lElemDtp = lDtp->subDTypep();
+                AstNodeDType* const rDtp = item.m_rhsp->dtypep();
+                if (lElemDtp->isSame(rDtp)) {
```

Comment:
Should it be isSame?  Isn't it OK if both are typedefs of same thing - might want a dtypep->skipRef() above.